File: scala/139.scala
object Solution {
    def wordBreak(s: String, wordDict: List[String]): Boolean = {
        val m = wordDict.foldLeft(Map[String, Boolean]()) {
            case (m, word) => m + (word -> true)
        }
        val dp = Array.fill(s.length())(false)
        for (l <- 0 until s.length(); r <- l+1 until s.length()) {
            if (m.contains(s.slice(l, r))) {
                if (l == 0 || dp(l)) dp(r) = true
            }
        }
        dp.last
    }
}

File: leetcodeJava/139.java
package leetcodeJava;

class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        var dp = new boolean[s.length()+1];
        var dict = new HashSet<String>();
        wordDict.stream().forEach(x -> dict.add(x));
        dp[0] = true;
        for (var i = 1; i <= s.length(); i++) {
            for(var j=0; j < i; j++) {
                if (dp[j] && dict.contains(s.substring(j, i))) {
                    dp[i] = true;
                }
            }
        }
        return dp[s.length()];
    }
}
